examples/pngWithRoundedShapes.php

```diff
@@ -0,0 +1,210 @@
+            imagefilledrectangle(
+                $this->image,
+                $x2 - $rectsize,
+                $y2 - $rectsize,
+                $x2,
+                $y2,
+                $this->moduleValues[$M_TYPE]
+            );
+        }
+
+        // ---------------------
+        // inner rounding
+        // ---------------------
+
+        if (!$this->matrix->check($x, $y)) {
+
+            if ($neighbours & 1 && $neighbours & (1 << 7) && $neighbours & (1 << 1)) {
+                // top left
+                imagefilledrectangle(
+                    $this->image,
+                    $x1,
+                    $y1,
+                    $x1 + $rectsize,
+                    $y1 + $rectsize,
+                    $this->moduleValues[$M_TYPE | QRMatrix::IS_DARK]
+                );
+            }
+
+            if ($neighbours & (1 << 1) && $neighbours & (1 << 2) && $neighbours & (1 << 3)) {
+                // top right
+                imagefilledrectangle(
+                    $this->image,
+                    $x2 - $rectsize,
+                    $y1,
+                    $x2,
+                    $y1 + $rectsize,
+                    $this->moduleValues[$M_TYPE | QRMatrix::IS_DARK]
+                );
+            }
+
+            if ($neighbours & (1 << 7) && $neighbours & (1 << 6) && $neighbours & (1 << 5)) {
+                // bottom left
+                imagefilledrectangle(
+                    $this->image,
+                    $x1,
+                    $y2 - $rectsize,
+                    $x1 + $rectsize,
+                    $y2,
+                    $this->moduleValues[$M_TYPE | QRMatrix::IS_DARK]
+                );
+            }
+
+            if ($neighbours & (1 << 3) && $neighbours & (1 << 4) && $neighbours & (1 << 5)) {
+                // bottom right
+                imagefilledrectangle(
+                    $this->image,
+                    $x2 - $rectsize,
+                    $y2 - $rectsize,
+                    $x2,
+                    $y2,
+                    $this->moduleValues[$M_TYPE | QRMatrix::IS_DARK]
+                );
+            }
+        }
+
+        imagefilledellipse(
+            $this->image,
+            (int)(($x * $this->scale) + ($this->scale / 2)),
+            (int)(($y * $this->scale) + ($this->scale / 2)),
+            (int)(2 * $this->options->circleRadius * $this->scale),
+            (int)(2 * $this->options->circleRadius * $this->scale),
+            $this->moduleValues[$M_TYPE]
+        );
+    }
+}
+
+
+// --------------------
+// Example
+// --------------------
+
+$options = new QROptions([
+    'outputType'=> QROutputInterface::CUSTOM,
+    'outputInterface' => QRCustomPNG::class,
+    'eccLevel' => EccLevel::M,
+    'imageTransparent' => false,
+    'imageBase64' => false,
+    'scale' => 20
+]);
+
+$qrcode = new QRCode($options);
+
+$img = $qrcode->render("https://www.youtube.com/watch?v=dQw4w9WgXcQ");
+
+header('Content-type: image/png');
+
+echo $img;
```

Comment:
Other than the mentioned it's mostly just code formatting, but don't worry about that, I'll just run it through the formatter later to align it with the other examples